[Tan Wen Cong] iP

TanWenCong opened the pull request:
# Duke
> Subtracting from your list of priorities is as important as adding to it. - Frank Sonnenberg, [Soul Food: Change Your Thinking, Change Your Life](https://www.goodreads.com/work/quotes/66520153)

Leave all the memory work to Duke! Duke is
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is 
1. Download it from [here](https://github.com/TanWenCong/ip)
2. Git clone
3. Run the jar file
4. One example command you can execute is `todo read` when running Duke

And it is **FREE**! :grin:

Features:
- [X] Managing Todo
- [X] Managing Deadline
- [X] Managing Event

Here is the main method, you may edit the file path to store your tasks at:

```java
public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

TanWenCong committed:
Skeletal version of Duke: greets user, echo user commands, exit if user types "bye"

TanWenCong committed:
Added 2 new features: Add user commands to history and List out history of user commands

TanWenCong committed:
Added new class Task. Added new feature Mark as Done

TanWenCong committed:
Added 3 children class of Task: Todo, Event, Deadline. Refactored function processUserCommand

TanWenCong committed:
Fixed bug where there is extra spacing when displaying event and deadline. Added testcases to semi-automate testing of Duke

TanWenCong committed:
Added custom Exception: DukeException. Invalid user inputs throws DukeExceptions.
Updated testcases for exceptions

TanWenCong committed:
Duke.java: Added function deleteTask, renamed function addToList to addTasks
EXPECTED.TXT, input.txt: updated testcases for deleteTask

TanWenCong committed:
Update EXPECTED.TXT
runtest.sh: pause results window before closing

TanWenCong committed:
Update Duke.java to follow Java coding standards

TanWenCong committed:
Extract Duke methods into OOP style

Duke.java contains main code features.

When more features are added, Duke.java becomes cluttered. This reduces
readability.

Methods are extracted into various java files (AddCommand... etc)

This reduces code length in Duke.java, and methods are allocated
appropriately in their respective new classes.

TanWenCong committed:
Merge branch 'master' into branch-Level-7

TanWenCong committed:
Merge branch 'master' into branch-Level-8

TanWenCong committed:
Add Storage.java

TaskList is always generated from scratch.

This does not allow retention of tasks created before.

Creating Storage saves the TaskList and allows loading when Duke is run
subsequently.

TanWenCong committed:
Add LocalDateTime to Deadline.java

Deadline stores date as String.

Storing date as LocalDate object is more meaningful. It gives access to
methods that deal with dates, such as comparing which date comes first.

TanWenCong committed:
Merge branch 'branch-Level-7'

TanWenCong committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Parser.java

TanWenCong committed:
Add all files to package duke

TanWenCong committed:
Merge pull request #1 from nus-cs2103-AY2223S1/add-gradle-support

[Tan Wen Cong] iP

TanWenCong committed:
Add DeadlineTest.java

Current Duke has no test automation. As program grows, it will become
harder to spot bugs.

Use Gradle and junit to help automate building and testing.

TanWenCong committed:
Add JavaDoc to all classes

TanWenCong committed:
Update code to meet coding standards

TanWenCong committed:
Add new feature Find

User stores tasks in taskList but can only choose to access all tasks.

This is not user-friendly. User might just want to look for relevant
tasks in the whole list.

User can now find tasks in taskList using command "find" followed by
a keyword to filter tasks containing that keyword.

TanWenCong committed:
Merge branch 'branch-Level-9' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java

Kept changes in both branches

TanWenCong committed:
Update JavaDoc for FindCommand.java and Task.java

TanWenCong committed:
Merge branch 'branch-Level-9'

TanWenCong committed:
Merge branch 'branch-A-CodingStandard'

TanWenCong committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Task.java

Kept changes made in branch-A-JavaDoc

TanWenCong committed:
Tagged A-Jar

hauchongtang commented on src/main/java/duke/AddCommand.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) {
+        taskList.addTask(task);
+        storage.save(taskList);
```
Good to see that you have abstracted parsing and task storage logic to `TaskList` and `Storage` classes.

hauchongtang commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,19 @@
+package duke;
+
+public abstract class Command {
```
Good use of abstract classes, since `Command` need not be instantiated anywhere in the code.

hauchongtang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+            return new Event(temp[0], temp[1].split(" ", 2)[1], isDone);
+        case "deadline":
+            return new Deadline(temp[0], temp[1].split(" ", 2)[1], isDone);
+        default:
+            // should not reach this part
+            return null;
```
Try not to return null here, you could instead throw an Exception, reason being that for possible future extensions of Duke, it would be harder to debug, especially when faced with `NullPointerException`.

hauchongtang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Scanner;
+
+public class Storage {
+    String filepath;
```
Would you need `filepath` in other classes ? Else, you should consider maintaining private accessibility.

hauchongtang commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,74 @@
+package duke;
+
+public abstract class Task {
```
Similar to `Command` class, good job in instantiating it as an abstract class.

hauchongtang commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Shows the tasks in TaskList without additional comments
+     */
+    private void showTasksPlain() {
```
Maybe you can consider renaming this function for better clarity based on what your code logic for this function is, which seems to be outputting a numbered list of tasks stored in .txt file locally.
```suggestion
    private void showNumberedTasks() {
```

hauchongtang reviewed (commented):
Overall, code was written well with good code quality. However, there are some minor stuff you should take note of. Also, it would be better if you could package your classes accordingly based on what role do they play in this chatbot for better readability and extensibility in future.

TanWenCong committed:
Add GUI for Duke using JavaFX

User interacts with Duke using Command Line Interface (CLI).

CLI is not visually appealing, user may not like CLI.

Using GUI allows us to beautify the process of using Duke, giving the
user a better experience.

Let's use JavaFX to implement GUI for Duke.

nareus commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+            throw new DukeException("I'm sorry, but I don't know what that means.");
+        }
+
+        switch (userCommand) {
```
I like that you have used indentation here appropriately

nareus commented on src/main/java/duke/UnmarkCommand.java:
```diff
@@ -0,0 +1,27 @@
+package duke;
+
+public class UnmarkCommand extends Command {
```
I like that the names of classes and methods are following coding standards

nareus reviewed (commented):
Looks good. Coding standard seems to be in order

TanWenCong committed:
Add Checkstyle script

Code runs fine, but we are not sure if all the code follows the coding
standard.

We need to follow coding standard to ensure clean code. Checkstyle can
help automate the process of finding code that does not follow the
standard.

Let's include Checkstyle to help automate this process.

TanWenCong committed:
Merge branch 'branch-Level-10'

TanWenCong committed:
Merge branch 'branch-A-Checkstyle'

TanWenCong committed:
Add assert statement in Parser.java

Parser parses user input and creates Command instance to execute.

Parser may have unexpected behavior and return null as Command.

Let's use assert to ensure Parser never returns a null Command.

TanWenCong committed:
Update code quality

Code is working, but code quality is not up to standard.

Bad code quality makes it hard to read and maintain the codebase.

Let's follow the checkstyle and coding standards to fix the code
quality.

TanWenCong committed:
Merge pull request #2 from TanWenCong/branch-A-Assertions

Add assert statement in Parser.java

TanWenCong committed:
Merge branch 'master' into branch-A-CodeQuality

TanWenCong committed:
Merge pull request #3 from TanWenCong/branch-A-CodeQuality

Update code quality

TanWenCong committed:
Update build.gradle to create Jar file

TanWenCong committed:
Add ArchiveCommand

Duke can save and load tasks from a single file.

User may want to archive tasks and start new on a clean state. We will
need more files to store archives.

Let's implement archive method in Storage and store archives in files
named by the time it was archived.

TanWenCong committed:
Set up Continuous Integration (CI)

TanWenCong committed:
Add files via upload

TanWenCong committed:
Add User Guide

TanWenCong committed:
Merge branch 'master' of https://github.com/TanWenCong/ip

TanWenCong committed:
Update Duke

Include greeting message to be shown on startup

TanWenCong committed:
Remove unused imports
